### ISSUE TITLE
phoneme timing が実音声に対し系統的に短い — ONNX durations は ceil 前の値（全ランタイム、spec 改訂を伴う）

### ISSUE THREAD

**Issue by ayutaz** [ayutaz/piper-plus#653]
## 概要

ONNX の `durations` 出力は `torch.ceil` 適用**前**の float (`w`)。一方で実際の音声フレーム割当は `w_ceil = ceil(w)` で決まる。現行の phoneme timing は raw `durations` をそのまま cursor walk する仕様（`docs/spec/phoneme-timing-contract.toml` 規定）のため、**timing が実音声より系統的に短くなる**。1 音素あたり平均 +0.5 frame の取りこぼしが累積するので、音素数に比例して誤差が拡大する。

全ランタイム（canonical の Python を含む）が spec 通りに実装しているため parity は保たれているが、**値そのものが実音声と一致しない**。

## 根拠（コード）

ONNX export 側 — `src/python/piper_train/export_onnx.py:619-623`:

```python
w = torch.exp(logw) * x_mask * length_scale
durations = w.squeeze(1)  # [batch, phoneme_length]  ← ONNX output（ceil 前）
...
w_ceil = torch.ceil(w)    # ← attention path / 実フレーム割当はこちら
```

timing 側 — raw durations をそのまま積算:

- spec: `docs/spec/phoneme-timing-contract.toml` `[calculation.cursor_walk]`
- Python (canonical): `src/python_run/piper_plus/timing.py:70-` (`cursor_ms += max(dur, 0) * frame_time_ms`)
- C++: `src/cpp/piper.cpp:365-398`

C++ はさらに、`start_frame` / `end_frame` が `static_cast<int>(duration)` の**切り捨て**で積算される一方、`start_time` / `end_time` は raw float 積算という内部不整合もある（`src/cpp/piper.cpp:375-376`, `397-398`）。

## 実測

dev `244ffeb4` / `test/models/multilingual-test-medium.onnx`（`durations` 出力あり）/ `--language es` / hop=256 / 22050Hz
text: `Hola, esta es una prueba.`（98 phoneme ids）

| 指標 | frames | 秒 |
|---|---|---|
| Σ raw durations | 98.06 | 1.138 |
| Σ ceil(durations) | 128 | 1.486 |
| 実出力音声 | 118 | 1.370 |
| timing 最終音素の end | 88.63 | 1.029 |

- `TRIM_EOS_MAX_FRAMES = 0`（`src/cpp/piper.cpp:99`）により EOS 区間 `ceil(d_EOS) = 10` frames が除去されるため **`118 = 128 - 10` と完全一致** → 実音声のフレーム配分が `ceil(d_i)` 側であることが確定。
- timing の終端は実音声に対し **-25%**（1.029s vs 1.370s）。長文では末尾ほどずれが拡大する。

## 影響

- lip-sync / 字幕 / 強制アライメント用途で、末尾ほど音とずれる（用途上致命的）
- 全 6 ランタイム共通（`docs/spec/phoneme-timing-contract.toml` `[implementations]` 記載の全実装）
- parity テストは現行仕様を pin しているため CI では検出されない

## 修正方針（要判断）

- **A. cursor を `ceil(d_i)` 基準に変更（推奨）** — 物理的に正しい。`docs/spec/phoneme-timing-contract.toml` の `spec_version` bump + 算式更新 + 全ランタイム修正 + parity fixture 再生成が必要。
- B. export 側で `durations = w_ceil` を出力 — 既存配布モデルとの非互換が発生するため非推奨。

## 受け入れ条件

- [ ] Σ(timing duration) が実出力音声長（EOS trim 前）と一致する
- [ ] 全ランタイムで byte-identical（parity fixture 再生成込み）
- [ ] `docs/spec/phoneme-timing-contract.toml` の `spec_version` bump と `[calculation]` 更新
- [ ] C++ の `start_frame`/`end_frame` と `start_time`/`end_time` が同一基準になる
- [ ] 挙動変更をマイグレーションガイドに記載

## 関連

#652 の調査中に発見。#652 は「C++ で timing が呼び出し元に伝播しない」別問題で、本 issue はその修正後に残る精度の問題。

**Comment by ayutaz** [comment]
## 調査結果 — 方針 A が厳密解であることを確認、ただし着手前に潰すべき前提が 4 件

### 1. `ceil(d_i)` が実フレーム配分と完全一致することを実測

`test/models/multilingual-test-medium.onnx` (hop=256 / 22050Hz / `--language es`) で 4 ケース測定:

| text | ids | Σraw | Σceil | 実出力 | raw/実 |
|---|---:|---:|---:|---:|---:|
| `Hola, esta es una prueba.` | 66 | 90.566 | **123** | **123** | 73.6% |
| `Hola.` | 16 | 26.685 | **35** | **35** | 76.2% |
| `Sol` | 10 | 20.527 | **27** | **27** | 76.0% |
| 長文 (208 ids) | 208 | 263.979 | **358** | **358** | 73.7% |

**4/4 で `Σceil == 実出力フレーム数`**。さらに `export_onnx.py:622-628` → `commons.generate_path()` が `cumsum(w_ceil)` の差分でパスを作ることをコード上で確認したので、音素 *i* が占めるフレーム区間は厳密に `[Σ_{j<i} ceil(d_j), Σ_{j≤i} ceil(d_j))` である。つまり方針 A は近似ではなく**厳密解**。

### 2. 前提: golden fixture が本変更を検出できない

`tests/fixtures/phoneme_timing/golden_matrix.json` の 7 ケースは**全て整数 durations** (最大 10 frame、最長 11 音素):

```
basic_konnichiwa   n=11 小数=0 max=8
single_phoneme     n= 1 小数=0 max=10
negative_clamped   n= 3 小数=0 max=5
high_sample_rate   n= 2 小数=0 max=10
pua_phoneme        n= 3 小数=0 max=8
empty              n= 0 小数=0 max=0
all_zero_durations n= 3 小数=0 max=0
```

`ceil(整数) == 整数` なので、`ceil` を 6 ランタイムに入れても fixture は **1 バイトも変わらず**、parity テストも全て緑のまま通る。実際の ONNX 出力は常に小数 (上表の 90.566 等) なので、fixture の入力はあらゆる実呼び出しを代表していない。**小数ケースと長文ケースの追加が先**。

### 3. 前提: C++ の parity テストが本番コードを 1 行も通っていない

`src/cpp/tests/test_phoneme_timing_parity.cpp:21` が明記している:

> To stay self-contained (no onnxruntime/espeak link), this test re-implements the spec algorithm

他 5 ランタイムは本番関数を呼んでいる (Python `durations_to_timing` / Rust `durations_to_timing` / Go `DurationsToTiming` / JS `durationsToTiming` / C# `TimingWriter.CalculateTiming`)。**C++ だけレプリカ**なので、C++ のテスト側を `ceil` にしても本番 `piper.cpp:369 extractTimingsFromDurations` は無変更のまま緑になる。

PR #687 で `trim_helpers.hpp` を切り出したのと同じ処置 (ORT 非依存ヘッダへの抽出) が必要。#703 に同型 5 箇所をまとめた。

### 4. 前提: cursor の型が 6 ランタイムで揃っていない

C++ (`piper.cpp:393 float currentTime`) と C# (`TimingWriter.cs:111 float currentTimeMs`) は **float32**、Python / Rust / Go / JS は float64。累積誤差を実測 (C、ceil 後の整数フレームを累積):

| 規模 | frames | f64 との差 |
|---|---:|---:|
| fixture 相当 (11 音素) | 60 | 6.8e-5 ms |
| 短文 (66 音素) | 397 | 8.4e-4 ms |
| 長文 (208 音素) | 1153 | **2.1e-3 ms** |

C++ の parity 許容誤差は `kAbsTolMs = 1e-6`、C# は `precision: 3` (1e-3 ms) と**そもそも揃っていない**。`ceil` で値が ~35% 増えるため差は広がる。cursor を double に統一し、許容誤差も揃える必要がある。

### 5. 前提: `timing-parity.yml` の trigger が 6 ランタイムを含んでいない

`paths:` は `src/python_run/piper_plus/timing.py` と fixture/spec/script のみ。他 5 ランタイムの timing 実装も、`per-runtime-presence` job が守っている parity テスト 6 件も入っていないため、それらだけを変更する PR では workflow が起動しない。#704 に同型 8 workflow をまとめた。

### 6. C++ の内部不整合 (issue 本文の指摘を確認)

`piper.cpp:402-425` で `currentFrame += static_cast<int>(duration)` (切り捨て) と `currentTime += duration * frameLength` (raw float) が併存。両方を `ceil` 基準に揃える。

### 着手順の提案

1. #703 の 2 番 (`extractTimingsFromDurations` をヘッダ抽出) — C++ を検査可能にする
2. #704 の `timing-parity.yml` trigger 拡張 — 6 ランタイムの変更で gate が起動するようにする
3. fixture に小数ケース・長文ケース・cursor 型が効くケースを追加 → この時点で `ceil` 無しの現行実装が **RED になることを確認**
4. cursor を double に統一、許容誤差を 6 ランタイムで揃える
5. `ceil` を 6 ランタイムに実装 → fixture 再生成 → GREEN
6. `spec_version` 1.1 → 1.2、`[calculation]` 更新、マイグレーションガイド追記
7. #697 (padding エントリ / `total_duration_ms` 超過) を同梱
